src/connectors/impls/gpubsub/producer.rs

```diff
@@ -0,0 +1,213 @@
+
+            Ok(Box::new(GpubConnector { config }))
+        } else {
+            Err(ErrorKind::MissingConfiguration(alias.to_string()).into())
+        }
+    }
+}
+
+struct GpubConnector {
+    config: Config,
+}
+
+#[async_trait::async_trait()]
+impl Connector for GpubConnector {
+    async fn create_sink(
+        &mut self,
+        sink_context: SinkContext,
+        builder: SinkManagerBuilder,
+    ) -> Result<Option<SinkAddr>> {
+        let sink = GpubSink {
+            config: self.config.clone(),
+            url: Url::parse(&self.config.endpoint)?,
+            client: None,
+        };
+        builder.spawn(sink, sink_context).map(Some)
+    }
+
+    async fn connect(&mut self, _ctx: &ConnectorContext, _attempt: &Attempt) -> Result<bool> {
+        Ok(true)
+    }
+
+    fn codec_requirements(&self) -> CodecReq {
+        CodecReq::Required
+    }
+}
+
+struct GpubSink {
+    config: Config,
+    url: Url<HttpsDefaults>,
+
+    client: Option<PublisherClient<InterceptedService<Channel, AuthInterceptor>>>,
+}
+
+#[async_trait::async_trait()]
+impl Sink for GpubSink {
+    async fn connect(&mut self, _ctx: &SinkContext, _attempt: &Attempt) -> Result<bool> {
+        let mut channel = Channel::from_shared(self.config.endpoint.clone())?
+            .connect_timeout(Duration::from_nanos(self.config.connect_timeout));
+        if self.url.scheme() == "https" {
+            let tls_config = ClientTlsConfig::new()
+                .ca_certificate(Certificate::from_pem(googapis::CERTIFICATES))
+                .domain_name(
+                    self.url
+                        .host_str()
+                        .ok_or_else(|| Status::unavailable("The endpoint is missing a hostname"))?
+                        .to_string(),
+                );
+
+            channel = channel.tls_config(tls_config)?;
+        }
+
+        let channel = channel.connect().await?;
+        let token = Token::new()?;
+
+        let client = PublisherClient::with_interceptor(
+            channel,
+            AuthInterceptor {
+                token: Box::new(move || {
+                    token.header_value().map_err(|_| {
+                        Status::unavailable("Failed to retrieve authentication token.")
+                    })
+                }),
+            },
+        );
+
+        self.client = Some(client);
+
+        Ok(true)
+    }
+
+    async fn on_event(
+        &mut self,
+        _input: &str,
+        event: Event,
+        ctx: &SinkContext,
+        serializer: &mut EventSerializer,
+        _start: u64,
+    ) -> Result<SinkReply> {
+        let client = self.client.as_mut().ok_or(ErrorKind::ClientNotAvailable(
+            "PubSub",
+            "The publisher is not connected",
+        ))?;
+
+        let mut messages = vec![];
+
+        for e in event.value_iter() {
+            messages.push(PubsubMessage {
+                data: serializer.serialize(e, event.ingest_ns)?[0].clone(),
```

Comment:
We don't allow `[]` access in hot code since it can panic

Reply:
Yeah. I think it is safe to concat all the vecs from a single value and put it into 1 PubSubMessage.

Reply:
Ja, I think that's what we do in kafka isn't it? It should probably act the same.

Reply:
Oh, dang, thank you for reminding me. What does it actually mean when there's more than 1 result from the serializer? It seems like Kafka is actually sending each as a separate message.

Reply:
Okay, sounds good, I'll do it this way.
BTW Unless I'm reading the code incorrectly, kafka will issue a separate message per payload: https://github.com/tremor-rs/tremor-runtime/blob/main/src/connectors/impls/kafka/producer.rs#L259-L294

Reply:
You are totally right, then we should probably keep it that way